Added displayUsername and username update functionality

icatat opened the pull request:
<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #35525 

 I added the `displayUsername` in the user schema, I manage to use that on the front end instead of the username (in case displayUsername exists for the specific user). Now I am working on updating the username on the profile settings. I manage to have the user change in username with the casing they are wishing for. The username is saved in the lower case and the displayUsername saves the original case. However, I noticed there are 2 functions that are structured and placed differently. There is a function called `User.prototype.updateMyUsername` in `api-server/common/models/user.js` on line 572, and there is an `updateMyUsername` function in `api-server/server/boot/settings.js` on line 189. For the change that I made, it seemed like the `settings.js` was doing the changes, but not too sure whether these 2 function are serving a separate scope or not.

I know that fcc is using loopback, and if I understood correctly, the changes that we are making to update the username are making a call to `/update-my-username` function and then to the `updateMyUsername` function in `settings.js` ( I might be wrong). However, as I am not too familiar with loopback, I just wanted to make sure that we are not modifying something that should be auto-generated by some function, so I just wanted to double check

Commit:
Added displayUsername and username update functionality

raisedadead commented on api-server/common/models/user.js:
```diff
@@ -331,7 +335,8 @@ export default function(User) {
 
   User.doesExist = function doesExist(username, email) {
-    if (!username && (!email || !isEmail(email))) {
+    username = username.toLowerCase();
```
Operate on the property only after the validation check in the next line. This would throw an exception if the property is not valid.

nikrb replied:
@raisedadead moved this down a statement

raisedadead commented on api-server/common/models/user.js:
```diff
@@ -376,6 +381,7 @@ export default function(User) {
   });
 
   User.about = function about(username, cb) {
+    username = username.toLowerCase();
```
Operate on the property only after the validation check in the next line. This would throw an exception if the property is not valid.

nikrb replied:
@raisedadead moved down again

raisedadead commented on api-server/common/models/user.js:
```diff
@@ -565,14 +571,15 @@ export default function(User) {
   };
 
   User.prototype.updateMyUsername = function updateMyUsername(newUsername) {
+    console.log(newUsername);
```
```suggestion
    
```
Just a reminder comment to remove this later, when you are ready with the PR.

nikrb replied:
@raisedadead removed console.log

raisedadead commented on api-server/server/boot/settings.js:
```diff
@@ -188,18 +188,19 @@ function updateMyAbout(req, res, next) {
 
 function createUpdateMyUsername(app) {
   const { User } = app.models;
+  console.log(User);
```
```suggestion
  
```
Just a reminder comment to remove this later, when you are ready with the PR.

nikrb replied:
@raisedadead removed console.log

raisedadead reviewed (changes requested):
Hi @icatat 

Sincerest apologies about the delay. This is some amazing work! I will pull this down and see if we can get this into master for our next major release.

I have left a few comments for minor stuff, other than that it would be great if you could throw in some tests to go with this.

Thanks a lot for all your work on this. It would be nice feature.

raisedadead commented:
Now, coming to your queries: 

> [...]  I noticed there are 2 functions that are structured and placed differently [...]

You are correct, these are redundant methods defined for achieving the same result. Just an artefact of our legacy code.

Not your fault at all, sorry about the confusion. However updating both methods is fine, it will help us avoid bugs when we cleanup the mess.

Yes, as for the API handlers, you are correct in modifying them. So essentially we are all good here.

raisedadead commented:
Bump.

nikrb committed:
Merge branch 'master' into fix/make-url-case-insensitive

nikrb committed:
fix: linting

formatting updated for prettier

nikrb committed:
fix: remove console debug

removed console.log

nikrb committed:
fix: remove console debug

removed console.log

nikrb committed:
fix: move username assignment to safe place

moved the username assignment down a statement so that it doesn't cause exception as per mrugesh's comment

nikrb commented:
@raisedadead I have built and run after pulling the updated PR but see a couple of problems.

1. I'm seeing a console warning for a few reactjs lifecycle methods which are no longer available
2. I'm not sure I'm in the correct field, but the username in the settings page after signing in as a dev user locally says an uncontrolled input is being changed to a controlled input.

If you can confirm the same issues (and I have my ducks in a row) I can have a look at fixes.

raisedadead commented:
> I'm seeing a console warning for a few reactjs lifecycle methods which are no longer available.

That should be fine. We do need to update a lot of these methods. You can ignore these warnings.

> I'm not sure I'm in the correct field, but the username in the settings page after signing in as a dev user locally says an uncontrolled input is being changed to a controlled input.

I think you would need to get you own pair of Auth0 keys and set the LOCAL_MOCK_AUTH key to false to get the real user, but I am also inclined that this should not be a need. Can you help me with a few screenshots to understand this better?

nikrb commented:
@raisedadead settings screen:
![usernamefieldsnapshot](https://user-images.githubusercontent.com/16709534/64651284-05515d00-d419-11e9-89e5-60e7aa5c6260.png)


and the warning text: 
```
commons.js:494536 Warning: A component is changing an uncontrolled input of type undefined to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component.
More info: https://fb.me/react-controlled-components
    in input (created by FormControl)
    in FormControl (created by UsernameSettings)
    in div (created by FormGroup)
    in FormGroup (created by UsernameSettings)
    in div (created by Col)
    in Col (created by FullWidthRow)
    in div (created by Row)
    in Row (created by FullWidthRow)
    in FullWidthRow (created by UsernameSettings)
    in form (created by UsernameSettings)
    in UsernameSettings (created by Connect(UsernameSettings))
    in Connect(UsernameSettings) (created by AboutSettings)
    in div (created by AboutSettings)
    in AboutSettings (created by ShowSettings)
    in main (created by ShowSettings)
```

raisedadead commented on api-server/common/models/user.js:
```diff
@@ -331,9 +335,10 @@ export default function(User) {
 
   User.doesExist = function doesExist(username, email) {
-    if (!username && (!email || !isEmail(email))) {
+    if (!username.toLowerCase() && (!email || !isEmail(email))) {
```
```suggestion
    if (!username && (!email || !isEmail(email))) {
```

You may not have the object itself first.

nikrb replied:
@raisedadead trying to commit your suggestion but seems I don't have permission to push to the repo. Can you commit them and @ojeytonwilliams 's suggestions?

ojeytonwilliams replied:
@nikrb have you tried pushing a new commit onto the branch?  Sometimes the Github web interface just doesn't work.

nikrb replied:
@ojeytonwilliams I tried a local push originally but it said I didn't have permission. Is that what you mean?

ojeytonwilliams replied:
Yep, you should have been able to push your local changes. I'll take a look when I can and try and see what's going on.

ojeytonwilliams commented on client/src/components/profile/components/Camper.js:
```diff
@@ -99,7 +101,12 @@ function Camper({
+      <h2 className='text-center username'>
+        @{displayUsername ? displayUsername : username}
+      </h2>
+      <h2 className='text-center username'>
```
```suggestion
```
Debugging code, should be removed.

ojeytonwilliams commented on client/src/components/profile/components/Camper.js:
```diff
@@ -99,7 +101,12 @@ function Camper({
+        @{displayUsername ? displayUsername : username}
+      </h2>
+      <h2 className='text-center username'>
+        @Username(for testing only): {username}
```
```suggestion
```
See above

ojeytonwilliams commented on client/src/components/profile/components/Camper.js:
```diff
@@ -99,7 +101,12 @@ function Camper({
+      </h2>
+      <h2 className='text-center username'>
+        @Username(for testing only): {username}
+      </h2>
```
```suggestion
```
See above.

ojeytonwilliams commented on client/src/components/settings/Username.js:
```diff
@@ -59,6 +60,7 @@ class UsernameSettings extends Component {
     this.state = {
       isFormPristine: true,
       formValue: props.username,
+      formDisplayValue: props.displayUsername,
```
```suggestion
      formDisplayValue: props.displayUsername
        ? props.displayUsername
        : props.username,
```
Needs to fallback to `username` in case `displayUsername` doesn't exist.

ojeytonwilliams reviewed (changes requested):
Hey, @nikrb, I tracked down the uncontrolled input problem - it happens when `displayUsername` isn't set.  I added a fallback as a suggestion that should fix it.

ojeytonwilliams committed:
fix: handle missing username or displayUsername

ojeytonwilliams commented:
@nikrb I'm not sure what was going wrong, but I managed to push @raisedadead and my suggestions.  That should stop the error you were seeing.

ojeytonwilliams reviewed (changes requested):
I had another look over this.  @nikrb could you check these suggestions are okay?

If they are and Github doesn't give you permission again, just let me know.

RandellDawson commented:
@raisedadead This PR has gone stale due to contributor not responding or fixing the issues @ojeytonwilliams pointed out above.

raisedadead commented:
I would still like to get this in. I am open if anyone is willing to close this PR and make a fresh one themselves taking all the propose code. Any takers?

ojeytonwilliams commented:
@raisedadead  I was thinking of just pushing the fixes onto this PR.  @icatat did do the heavy lifting, after all. 

I'm happy to finish it, either way.

raisedadead commented:
Thanks @ojeytonwilliams for taking this up. I recommend we squash everything to one single commit and push to this PR.

ojeytonwilliams committed:
Merge branch 'master' into fix/make-url-case-insensitive

ojeytonwilliams committed:
refactor: remove redundant code

ojeytonwilliams commented:
@raisedadead I'll squash it down before merging.  Did you want to take another look before I do that?

raisedadead commented:
Nope we should be OK once you have your suggestions in and resolved the conflicts (if any).

ojeytonwilliams commented:
@icatat Thank you for your contribution! :+1:
We're happy to accept these changes, and look forward to future contributions. :memo:
